tests/test_build_nanoset_dataloader.py

```diff
@@ -0,0 +1,130 @@
+import random
+
+import pytest
+from helpers.context import TestContext
+from helpers.data import (
+    assert_batch_dataloader,
+    create_dataset_paths,
+    create_dummy_json_dataset,
+    get_max_value_by_group,
+    preprocess_dummy_dataset,
+)
+from helpers.utils import available_gpus, get_all_3d_configurations, init_distributed, rerun_if_address_is_in_use
+from nanotron.data.blended_nanoset import BlendedNanoset
+from nanotron.data.dataloader_builder import build_nanoset_dataloader
+from nanotron.data.dataset_builder import NanosetBuilder
+from nanotron.data.nanoset import Nanoset
+from nanotron.data.nanoset_configs import NanosetConfig
+from nanotron.data.utils import compute_datasets_num_samples
+from nanotron.parallel import ParallelContext
+
+
+@pytest.mark.parametrize(
+    "tp,dp,pp",
+    [
+        pytest.param(*all_3d_configs)
+        for gpus in range(1, min(available_gpus(), 4) + 1)
+        for all_3d_configs in get_all_3d_configurations(gpus)
+    ],
+)
+@rerun_if_address_is_in_use()
```

Comment:
We should add a separate test to verify that, given the same seed, NanoSet is deterministic while also varying sequence_length, the total number of samples, split ratio... (yes, I am aware that currently in nanotron we also don't have these tests for the current dataloader, but from now on, we aim to rectify this by adding as many good tests as possible)

Reply:
Indeed, the three variables you mentioned are the ones that most influence the creation of the `Nanoset`. It is worth noting that when preprocessing the JSON file, what we do is tokenize each document (1 line of the JSON = 1 document) and concatenate all the tokens. That's why...

- When varying the sequence length, the samples you construct in the [`sample_index`](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L212) are different because what you actually do is take sequence length tokens + 1 consecutively from the `document_index` documents.
- When varying the number of samples, you increase or decrease the [number of times](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L234) you concatenate `shuffle_index` so that it is always greater than the number of samples and does not end the dataset.
- When varying the split, you change the number of documents used to create the Nanoset, so the result of [shuffling the `document_index`](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L196) and the quantity of samples (and the samples itself) in `sample_index` will be different.

It is important to remember that EVERYTHING random, both the shuffling of the [`document_index](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L196) and the [`shuffle_index`](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L231), is done with a `numpy.random.RandomState` and the seed.

Reply:
> When varying the sequence length, the samples you construct in the [sample_index](https://github.com/TJ-Solergibert/nanotron/blob/16f44d1b8177b38a7ae3a2322a912562dbdc244a/src/nanotron/data/nanoset.py#L212) are different because what you actually do is take sequence length tokens + 1 consecutively from the document_index documents.

Yep, I mean, given the same set of sequence length, split ratio, the total number of samples, and also the same seed, then we expect that NanoSet is deterministic

Reply:
Okey! [Done](https://github.com/TJ-Solergibert/nanotron/blob/f163c1d6574b5f8f1d7e9d51caa9f006ec35660a/tests/helpers/data.py#L119)!

We check all the important features and we extract a sample from the dataset (sample with index 23).